### PR TITLE
[CARBONDATA-1698]Adding support for table level compaction configuration

### PR DESCRIPTION
**Pull request by Xaprice**
Adding support for table level compaction configuration

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1489/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1907/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1908/

**Comment by chenliang613** [comment]
retest this please

**Comment by chenliang613** [comment]
@Xaprice please add the checklist as per the below template:

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 1.6, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/393/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1523/

**Comment by chenliang613** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1524/

**Comment by jackylk** [comment]
Can you describe how this feature can be used by user? What SQL syntax?

**Comment by Xaprice** [comment]
Be sure to do all of the following checklist to help us incorporate
your contribution quickly and easily:

- [x] Any interfaces changed?
      **no**
- [x] Any backward compatibility impacted?
      **no**
- [x]  Document update required?
      **Yes, data-management-on-carbondata.md has been updated**
- [x] Testing done
    Please provide details on
    - Whether new unit test cases have been added or why no new tests are required?
        **new unit test cases added**
    - How it is tested? Please attach test report.
        **unit test and tested on cluster with 7 nodes**
    - Is it a performance related change? Please attach the performance test report.
        **no**
    - Any additional information to help reviewers in testing this change.
        **no**

- [x] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.
    **NOT RELATED**

**Comment by Xaprice** [comment]
@jackylk 
User can create table by SQL below:
```
 CREATE TABLE tableWithCompactionOptions(
     intField INT,
     stringField STRING
     )
 STORED BY 'carbondata'
 TBLPROPERTIES('MAJOR_COMPACTION_SIZE'='10240',
 'AUTO_LOAD_MERGE'='true',
 'COMPACTION_LEVEL_THRESHOLD'='5,6',
 'COMPACTION_PRESERVE_SEGMENTS'='10',
 'ALLOWED_COMPACTION_DAYS'='5')
```
Thus user can specify compaction configurations in table level. The configurations are all optional, if not specified, corresponding configurations in carbon.properties will be used. Related document has been updated.

**Comment by Xaprice** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1616/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2009/

File: core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java [hunk 1]
```diff
@@ -863,6 +863,16 @@
   public static final String TABLE_BLOCKSIZE = "table_blocksize";
   // set in column level to disable inverted index
   public static final String NO_INVERTED_INDEX = "no_inverted_index";
+  // table property name of major compaction size
+  public static final String TBL_PROP_MAJOR_COMPACTION_SIZE = "major_compaction_size";
```
Comment (chenliang613) [inline core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java]: suggest keeping the consistent name format , how about MAJOR_COMPACTION_SIZE,  remove "TBL_PROP"

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1672/

File: core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java (hunk 1 above)
Comment (Xaprice) [reply core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java]: TBL_PROPs removed

**Comment by Xaprice** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/399/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1673/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2060/

**Comment by Xaprice** [comment]
fix code style problem, retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/430/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2086/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/431/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1701/

**Comment by Xaprice** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/446/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/447/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2097/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1718/

**Comment by jackylk** [comment]
Thanks for working on this, I will review it this weekend

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2330/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/791/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2025/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2343/

**Comment by Xaprice** [comment]
retest this please

**Comment by Xaprice** [comment]
Hi @chenliang613 , can you please take a look?

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2346/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/803/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2034/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/807/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2038/

**Comment by chenliang613** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2054/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/827/

**Comment by chenliang613** [comment]
retest this please

File: core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java [hunk 2]
```diff
@@ -873,6 +873,16 @@
   public static final String TABLE_BLOCKSIZE = "table_blocksize";
   // set in column level to disable inverted index
   public static final String NO_INVERTED_INDEX = "no_inverted_index";
+  // table property name of major compaction size
+  public static final String MAJOR_COMPACTION_SIZE = "major_compaction_size";
```
Comment (chenliang613) [inline core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java]: For these compaction properties of Table level , suggest adding "TABLE", such as : TABLE_MAJOR_COMPACTION_SIZE,TABLE_AUTO_LOAD_MERGE...

**Comment by chenliang613** [comment]
@Xaprice  Thanks for your contributions.   Overall changes look fine for me.

one question:  if user set the different compaction configuration for "Carbon system" and "Table level", should be using Table level's property ,is it right ?  can you add this part test cases ?

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/843/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2069/

**Comment by chenliang613** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2092/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/869/

**Comment by chenliang613** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/876/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2101/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/880/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2105/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2182/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/953/

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/datacompaction/TableLevelCompactionParameterTest.scala [hunk 3]
```diff
@@ -0,0 +1,183 @@
+/*
```
Comment (chenliang613) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/datacompaction/TableLevelCompactionParameterTest.scala]: This test case is same as TableLevelCompactionOptionTest.scala, please remove this one(TableLevelCompactionParameterTest.scala)

File: integration/spark2/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala [hunk 4]
```diff
@@ -205,7 +205,8 @@ object CarbonDataRDDFactory {
 
               val newCarbonLoadModel = prepareCarbonLoadModel(table)
 
-              val compactionSize = CarbonDataMergerUtil.getCompactionSize(CompactionType.MAJOR)
+              val compactionSize = CarbonDataMergerUtil
+                .getCompactionSize(CompactionType.MAJOR, carbonLoadModel)
```
Comment (chenliang613) [inline integration/spark2/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala]: why need add one more input parameter(carbonLoadModel) for getCompactionSize()?

Comment (Xaprice) [reply integration/spark2/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala]: carbonLoadModel may contain table-level major compaction size if it is specified in create table SQL, so the purpose for adding parameter 'carbonLoadModel' is to get the table-level major compaction size.

Comment (chenliang613) [reply integration/spark2/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala]: ok

**Comment by Xaprice** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/978/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2469/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2208/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/986/

**Comment by chenliang613** [comment]
LGTM, good contribution.

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2474/

**Pull request closed without merging**